README.txt

		soft_wdt软件看门狗简介
		
======================

    soft_wdt(以下简称本软件)是一个软件实现的Linux看门狗，
    他和/drivers/watchdog/softdog.c实现的软件看门狗几乎一样。
    主要的不同点是，前者支持一个看门狗，本软件则支持大量的看门狗。

    soft_wdt代码编译后，生成一个内核模块soft_wdt.ko。

    模块加载后，将创建一个设备文件/dev/soft_wdt

    用户态程序，通过系统调用open每打开一次/dev/soft_wdt，就得到一个新的看门狗，
    此看门狗的使用方法就和普通的看门狗一样。

    例如:
    1) 向fd写入任何数据，就等于是喂狗。
    2) 用户可以通过ioctl对看门狗进行各种操作。
    3) 如果模拟加载时，模块参数nowayout的值为0，
       那么当用户向fd写入一次含有字符V(注意，是大写)的数据时，
       就将此看门狗设置成了可关闭的。

下面具体说说使用方法：

(一)编译看门狗
    见Build.txt

(二)加载看门狗

    本软件提供的模块参数如下。用户可根据需要进行指定。

    nowayout           - 一旦启动看门狗，不可以停止 (0，no；1，yes。default=0)
    timeout            - 看门狗超时时间，单位：秒。 (0 ~ 65536, default=5)
    no_reboot          - 看门狗超时，不重启系统 。(0，no; 1，yes  default=0)
    core_dump_ill_task - 看门狗超时时，core dump异常任务，(0，no; 1，yes  default=1)

注意，core dump是通过向异常线程发送SIGABRT信号实现的。
因此，如果使用看门狗的程序，想自己记录异常信息，可以通过捕获SIGABRT信号来实现。

    下面是加载命令的示例。

    1. 使用默认参数加载(默认值如上面所列)
    insmod soft_wdt.ko

    2. 指定参数加载(12秒超时，看门狗可关闭，超时不重启机器)
    insmod soft_wdt.ko timeout=12 nowayout=0 no_reboot=1

(三)用户态程序使用看门狗

//以下是示例代码

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/watchdog.h>

#define    SOFT_WDT_DEV    "/dev/soft_wdt"

int feed_dog_cnt;
int feed_dog_gap;

int main(int argc, char *argv[])
{
    int i;
    int  timeout;
    struct watchdog_info ident;
    
    int fd;

    if (argc<3)
    {
        printf("usage:\n %s  <feed_gap(in seconds)>  <feed_cnt>\n", argv[0]);
        return 0;
    }
    
    fd=open("/dev/soft_wdt", O_WRONLY);

    if (fd < 0)
    {
        printf("open %s failed\n", SOFT_WDT_DEV);
        exit(1);
    }


    printf("open %s succeed\n", SOFT_WDT_DEV);
    
    timeout = 7;
    printf("set timeout to %d\n", timeout);
    ioctl(fd, WDIOC_SETTIMEOUT, &timeout);

    timeout = 0;
    ioctl(fd, WDIOC_GETTIMEOUT, &timeout);
    printf("get timeout returns %d\n", timeout);

    ioctl(fd, WDIOC_GETSUPPORT, &ident);
    printf("dog name is %s\n", ident.identity);

    printf("make the dog closable\n");
    write(fd, "V", 1);

    feed_dog_gap = atoi(argv[1]);
    feed_dog_cnt = atoi(argv[2]);
    for (i=0; i<feed_dog_cnt; i++)
    {
        printf("feed dog\n");
        write(fd, "1234", 4);
        usleep(feed_dog_gap*1000000);
    }

    printf("stop feeding dog\n");
    while (1)
    {
        usleep(1000000);
    }
    
    return 0;
}



======================

    本软件是一款开源、免费软件。
    具体版权说明见COPYING.txt。

======================

主要的版本历史：

2014年11月，1.0发布
2015年03月，2.0发布
====================== 
作者: 孙明保(来自 ZTE中兴)
邮箱: sunmingbao@126.com
QQ  : 7743896

release_notes.txt

	soft_wdt v2.0.3

发布日期：2022/03/16

【更新内容】
     o support more kernel versions.


soft_wdt_user.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define    SOFT_WDT_DEV    "/dev/soft_wdt"

int feed_dog_cnt;
int feed_dog_gap;

int main(int argc, char *argv[])
{
    int i;
    int fd=open("/dev/soft_wdt", O_WRONLY);

    if (fd < 0)
    {
        printf("open %s failed\n", SOFT_WDT_DEV);
        exit(1);
    }


    printf("open %s succeed\n", SOFT_WDT_DEV);

    write(fd, "<name>my_dog</name>", strlen("<name>my_dog</name>"));
    write(fd, "<timeout>123</timeout>", strlen("<timeout>123</timeout>"));
    write(fd, "<stop_on_fd_close>1</stop_on_fd_close>", strlen("<stop_on_fd_close>1</stop_on_fd_close>"));
    write(fd, "<no_reboot>1</no_reboot>", strlen("<no_reboot>1</no_reboot>"));
    
    feed_dog_gap = atoi(argv[1]);
    feed_dog_cnt = atoi(argv[2]);
    for (i=0; i<feed_dog_cnt; i++)
    {
        write(fd, "1234", 4);
        usleep(feed_dog_gap*1000000);
    }

    write(fd, "<stop_dog>1</stop_dog>", strlen("<stop_dog>1</stop_dog>"));
    while (1)
    {
        usleep(1000000);
    }
    
    return 0;
}


soft_wdt.c
/*
 * software watchdog timer, support multiple dogs
 *
 * Copyright (C) 2014-2015 Sun Mingbao <sunmingbao@126.com>
 * Dual licensed under the MIT and/or GPL licenses.
 *
 *	Implementation mechanism mainly learned from:
 *	SoftDog	0.07:	A Software Watchdog Device
 *
 *	(c) Copyright 1996 Alan Cox <alan@lxorguk.ukuu.org.uk>,
 *	                        All Rights Reserved.
 */

#include <linux/version.h>
#include <linux/init.h>
#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/kernel.h>
#include <linux/fs.h>
#include <linux/slab.h>
#include <linux/miscdevice.h>
#include <linux/watchdog.h>
#include <linux/reboot.h>
#include <linux/jiffies.h>
#include <linux/spinlock.h>
#include <linux/uaccess.h>
#include <linux/sched.h>
#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4,10,0))
#include <linux/signal.h>
#else
#include <linux/sched/signal.h>
#endif

#define MODULE_NAME	"soft_wdt"
#define MOD_VERSION	"2.0.3"
#define PFX		MODULE_NAME": "

/* time out in seconds */
#define TIMEOUT_MIN		(1)
#define TIMEOUT_MAX		(65536)
#define TIMEOUT_DEFAULT		(5)
#define TIMEOUT_VALID(v)	((v)>=TIMEOUT_MIN && (v)<=TIMEOUT_MAX)

/* module params */
static bool nowayout = WATCHDOG_NOWAYOUT;
module_param(nowayout, bool, 0);
MODULE_PARM_DESC(nowayout,
	"Watchdog cannot be stopped once started (default="
	__MODULE_STRING(WATCHDOG_NOWAYOUT) ")");

static int timeout = TIMEOUT_DEFAULT;
module_param(timeout, int, 0);
MODULE_PARM_DESC(timeout,
	"Watchdog timeout in seconds. (0 < timeout < 65536, default="
	__MODULE_STRING(TIMEOUT_DEFAULT) ")");

static bool no_reboot = 0;
module_param(no_reboot, bool, 0);
MODULE_PARM_DESC(no_reboot,
	"don't reboot the system on dog expire. (default=0)");

static bool core_dump_ill_task = 1;
module_param(core_dump_ill_task, bool, 0);
MODULE_PARM_DESC(core_dump_ill_task,
	"core dump the ill task on dog expire. (default=1)");


/* dog num */
static int max_dog_cnt = 128;
static int dog_cnt;
struct dog_struct {
	spinlock_t lock;
	struct list_head link_all;
	struct task_struct *owner;
	int id;
	int status;
	char expect_close;
	int is_orphan;
	struct watchdog_info ident;
	int timeout;
	struct timer_list ticktock;
};


#define DOG_NAME(dog)		((dog)->ident.identity)
#define DOG_SET_ALIVE(dog)	((dog)->status=0)
#define DOG_SET_NON_ALIVE(dog)	((dog)->status=1)
#define DOG_ALIVE(dog)		(0==(dog)->status)

static struct list_head all_dogs;
static spinlock_t all_dog_lock;

static void* find_dog_by_id(int id)
{
	struct dog_struct *dog;
	struct list_head *pos;
	list_for_each(pos, &all_dogs) {
		dog = list_entry(pos, struct dog_struct, link_all);
		if (dog->id == id)
			return dog;
	}
	return NULL;
}

static u16 generate_dog_id(void)
{
	static u16 next_id = 0;
	int ret;

	while (find_dog_by_id(next_id)) {
		next_id = (next_id + 1) % max_dog_cnt;
	}
	ret = next_id;
	next_id = (next_id + 1) % max_dog_cnt;
	return ret;
}

#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4,14,0))
static void dog_timeout(unsigned long data)
{	
	struct dog_struct *dog = (void *)data;
	
#else
static void dog_timeout(struct timer_list *t)
{
	struct dog_struct *dog = from_timer(dog, t, ticktock);
#endif
	char   dog_name[32];

	spin_lock(&dog->lock);
	if (!DOG_ALIVE(dog)) {
		spin_unlock(&dog->lock);
		return;
	}

	DOG_SET_NON_ALIVE(dog);
	spin_unlock(&dog->lock);

	/*
	 * for race condition that
	 * core_dump cause soft_wdt_release releases the dog
	 *
	 */
	strncpy(dog_name, DOG_NAME(dog), sizeof(dog_name));

	printk(KERN_WARNING PFX "%s expired.", DOG_NAME(dog));
	if (core_dump_ill_task && !(dog->is_orphan)) {
		printk(KERN_WARNING PFX "%s core dump ill task.",
			DOG_NAME(dog));
		send_sig(SIGABRT, dog->owner, 0);
	}

	if (!no_reboot) {
		printk(KERN_WARNING PFX "%s restart the system.",
			dog_name);
		emergency_restart();
	}

}

static void init_dog(struct dog_struct *dog)
{
	DOG_SET_ALIVE(dog);
	spin_lock_init(&(dog->lock));
	dog->owner = current;
	dog->is_orphan = 0;
	dog->timeout = timeout;
	dog->expect_close = 0;
	dog->ticktock =
#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4,14,0))
		(struct timer_list) TIMER_INITIALIZER(dog_timeout,
			0, (unsigned long)dog);
#else
		(struct timer_list) __TIMER_INITIALIZER(dog_timeout,TIMER_IRQSAFE);

#endif
	dog->ident.options =
		(WDIOF_SETTIMEOUT | WDIOF_KEEPALIVEPING | WDIOF_MAGICCLOSE);
	dog->ident.firmware_version = 0;
}


/*
 * operations around dog
 */
static void del_dog(struct dog_struct *dog)
{
	spin_lock(&all_dog_lock);
	list_del(&(dog->link_all));
	dog_cnt--;
	spin_unlock(&all_dog_lock);
}

static int feed_dog(struct dog_struct *dog)
{
	int ret = 0;
	spin_lock(&dog->lock);
	if (DOG_ALIVE(dog)) {
		mod_timer(&(dog->ticktock),
			jiffies + (dog->timeout * HZ));
	} else {
		printk(KERN_WARNING PFX "%s non-alive. invalid feed.",
			DOG_NAME(dog));
		ret = -EIO;
	}
	spin_unlock(&dog->lock);
	return ret;
}

static int add_dog(struct dog_struct *dog)
{
	init_dog(dog);
	spin_lock(&all_dog_lock);
	if (dog_cnt >= max_dog_cnt) {
		spin_unlock(&all_dog_lock);
		return -EUSERS;
	}

	dog->id = generate_dog_id();
	snprintf(DOG_NAME(dog), sizeof (DOG_NAME(dog)),
		"soft_wdt%d", dog->id);
	list_add(&(dog->link_all), &all_dogs);
	dog_cnt++;
	spin_unlock(&all_dog_lock);

	/* start the new dog */
	mod_timer(&(dog->ticktock), jiffies + (dog->timeout * HZ));
	return 0;
}

static int set_dog_timeout(struct dog_struct *dog, int t)
{
	if (!TIMEOUT_VALID(t)) {
		printk(KERN_WARNING PFX "set timeout failed."
			" timeout value %d not between [%d, %d].",
			t, TIMEOUT_MIN, TIMEOUT_MAX);
		return -EINVAL;
	}
	spin_lock(&dog->lock);
	dog->timeout = t;
	mod_timer(&(dog->ticktock), jiffies + (dog->timeout * HZ));
	spin_unlock(&dog->lock);
	printk(KERN_NOTICE PFX "%s timeout set to %d seconds.",
	DOG_NAME(dog), dog->timeout);
	return 0;
}

static void stop_dog(struct dog_struct *dog)
{
	spin_lock(&dog->lock);
	if (!DOG_ALIVE(dog)) {
		spin_unlock(&dog->lock);
		return;
	}

	del_timer(&(dog->ticktock));
	DOG_SET_NON_ALIVE(dog);
	spin_unlock(&dog->lock);
	printk(KERN_NOTICE PFX "%s stopped", DOG_NAME(dog));
}

/*
 * stop all dogs on system down
 */
static int soft_wdt_notify_sys(struct notifier_block *this,
					unsigned long code, void *unused)
{
	struct list_head *head = &all_dogs, *pos;
	struct dog_struct *dog;

	if (code == SYS_DOWN || code == SYS_HALT) {
		spin_lock(&all_dog_lock);
		list_for_each(pos, head) {
			dog = list_entry(pos, struct dog_struct, link_all);
			//pos = pos->prev;
			stop_dog(dog);
		}
		spin_unlock(&all_dog_lock);
	}

	return NOTIFY_DONE;
}

/*
 *	file operations
 */

static long soft_wdt_ioctl(struct file *file, unsigned int cmd,
				unsigned long arg)
{
	void __user *argp = (void __user *) arg;
	int __user *p = argp;
	struct dog_struct *dog = file->private_data;
	int timeout_new;

	switch (cmd) {
	case WDIOC_GETSUPPORT:
		return copy_to_user(argp, &(dog->ident),
				     sizeof (dog->ident)) ? -EFAULT : 0;
	case WDIOC_GETSTATUS:
	case WDIOC_GETBOOTSTATUS:
		return put_user(dog->status, p);

	case WDIOC_KEEPALIVE:
		feed_dog(dog);
		return 0;

	case WDIOC_SETTIMEOUT:
		if (get_user(timeout_new, p))
			return -EFAULT;
		if (set_dog_timeout(dog, timeout_new))
			return -EINVAL;

	case WDIOC_GETTIMEOUT:
		return put_user(dog->timeout, p);

	default:
		return -ENOTTY;
	}
}

static int soft_wdt_open(struct inode *inode, struct file *file)
{
	int ret = 0;
	struct dog_struct *dog;
	dog = kmalloc(sizeof(struct dog_struct), GFP_ATOMIC);
	if (NULL == dog)
		return -ENOMEM;

	ret = add_dog(dog);
	if (ret) {
		kfree(dog);
		return ret;
	}

	file->private_data = dog;
	nonseekable_open(inode, file);
	printk(KERN_NOTICE PFX "%s created.", DOG_NAME(dog));
	return 0;
}

static int soft_wdt_release(struct inode *inode, struct file *file)
{
	struct dog_struct *dog = file->private_data;

	dog->is_orphan = 1;

	if (dog->expect_close == 42 || !DOG_ALIVE(dog)) {
		stop_dog(dog);
		printk(KERN_NOTICE PFX "%s released.", DOG_NAME(dog));
		del_dog(dog);
		kfree(dog);

	}

	return 0;
}

static ssize_t soft_wdt_write(struct file *file, const char __user *data,
						size_t len, loff_t *ppos)
{
	struct dog_struct *dog = file->private_data;
	size_t i;
	char c;

	if (!len)
		goto OUT;

	feed_dog(dog);

	if (nowayout || dog->expect_close != 42)
		goto OUT;

	for (i = 0; i != len; i++) {
		if (get_user(c, data + i))
			return -EFAULT;
		if (c == 'V') {
			dog->expect_close = 42;
			printk(KERN_NOTICE PFX "%s expect close.",
						DOG_NAME(dog));
			goto OUT;
		}
	}

OUT:
	return len;
}


/*
 *	Kernel Interfaces
 */

static const struct file_operations soft_wdt_fops = {
	.owner		= THIS_MODULE,
	.llseek		= no_llseek,
	.write		= soft_wdt_write,
	.unlocked_ioctl	= soft_wdt_ioctl,
	.open		= soft_wdt_open,
	.release	= soft_wdt_release,
};

static struct miscdevice soft_wdt_miscdev = {
	.minor	= MISC_DYNAMIC_MINOR,
	.name	= MODULE_NAME,
	.fops	= &soft_wdt_fops,
};

static struct notifier_block soft_wdt_notifier = {
	.notifier_call	= soft_wdt_notify_sys,
};

static void __init init_inner_structure(void)
{
	dog_cnt = 0;
	spin_lock_init(&all_dog_lock);
	INIT_LIST_HEAD(&all_dogs);
}

static char banner[] __initdata =
	KERN_INFO "software watchdog timer, version: " MOD_VERSION
	" (nowayout=%d; timeout=%d; no_reboot=%d; core_dump_ill_task=%d)\n";

static int __init soft_wdt_init(void)
{
	int ret;
    
	if (!TIMEOUT_VALID(timeout)) {
		printk(KERN_WARNING PFX
			"timeout value %d not between [%d, %d]."
			"using default value %d",
			timeout, TIMEOUT_MIN, TIMEOUT_MAX, TIMEOUT_DEFAULT);
		timeout = TIMEOUT_DEFAULT;
	}
    
	init_inner_structure();
    
	ret = register_reboot_notifier(&soft_wdt_notifier);
	if (ret) {
		printk(KERN_ERR PFX
			"cannot register reboot notifier (err=%d)\n", ret);
		goto OUT;
	}

	ret = misc_register(&soft_wdt_miscdev);
	if (ret) {
		printk(KERN_ERR PFX "cannot register soft_wdt_miscdev");
		unregister_reboot_notifier(&soft_wdt_notifier);
		goto OUT;
	}

	printk(banner, nowayout, timeout, no_reboot, core_dump_ill_task);

OUT:
	return ret;

}

static void __exit deinit_inner_structure(void)
{
	struct list_head *head = &all_dogs, *pos;
	struct dog_struct *dog;
	spin_lock(&all_dog_lock);

	list_for_each(pos, head) {
		dog = list_entry(pos, struct dog_struct, link_all);
		pos = pos->prev;
		stop_dog(dog);
		printk(KERN_NOTICE PFX "%s released.", DOG_NAME(dog));
		list_del(&(dog->link_all));
		kfree(dog);
	}
	spin_unlock(&all_dog_lock);
}

static void __exit soft_wdt_exit(void)
{
	printk(KERN_INFO PFX "soft wdt quit");
	misc_deregister(&soft_wdt_miscdev);
	unregister_reboot_notifier(&soft_wdt_notifier);
	deinit_inner_structure();
}

module_init(soft_wdt_init);
module_exit(soft_wdt_exit);

MODULE_AUTHOR("Sun Mingbao <sunmingbao@126.com>");
MODULE_DESCRIPTION("software watchdog timer(support multiple dogs)");
MODULE_VERSION(MOD_VERSION);
MODULE_LICENSE("Dual MIT/GPL");
